storage/database.go

```diff
@@ -203,25 +205,67 @@ func (d *db) Write(
 	return shard.Write(ctx, id, timestamp, value, unit, annotation)
 }
 
-func (d *db) ReadEncoded(
-	ctx context.Context,
-	id string,
-	start, end time.Time,
-) ([][]xio.SegmentReader, error) {
+func (d *db) readableShard(shardID uint32) (databaseShard, error) {
 	d.RLock()
 	if !d.bsm.IsBootstrapped() {
 		d.RUnlock()
 		return nil, errDatabaseNotBootstrapped
 	}
-	shardID := d.shardSet.Shard(id)
 	shard := d.shards[shardID]
 	d.RUnlock()
 	if shard == nil {
 		return nil, fmt.Errorf("not responsible for shard %d", shardID)
 	}
+	return shard, nil
+}
+
+func (d *db) ReadEncoded(
+	ctx context.Context,
+	id string,
+	start, end time.Time,
+) ([][]xio.SegmentReader, error) {
+	shardID := d.shardSet.Shard(id)
+	shard, err := d.readableShard(shardID)
+	if err != nil {
+		return nil, err
+	}
 	return shard.ReadEncoded(ctx, id, start, end)
 }
 
+func (d *db) FetchBlocks(
+	ctx context.Context,
+	id string,
+	starts []time.Time,
+) []FetchBlockResult {
+	shardID := d.shardSet.Shard(id)
+	shard, err := d.readableShard(shardID)
+	if err != nil {
+		// If we don't own the shard associated with the id, return an error
+		// so the client knows it needs to retry these.
```

Comment:
For this case it's actually fine to just return a top level `xerrors.NewInvalidParamsError(...)`.  I perform error checking first at the top level and secondary at the block level if there are partial success and partial errors.  For entire request failed I'd much prefer to get just a full top level error (not wrapped in the result/response object).